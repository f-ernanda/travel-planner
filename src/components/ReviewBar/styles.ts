'use client'

import styled, { css } from 'styled-components'

export const Wrapper = styled.div`
  ${({ theme }) => css`
    display: flex;
    justify-content: center;

    > * {
      color: ${theme.colors.primary};
      flex: 0 0 100%;
      max-width: 4.8rem;
    }
  `}
`
