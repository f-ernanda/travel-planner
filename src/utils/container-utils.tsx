'use client'

import styled, { css } from 'styled-components'
import media from 'styled-media-query'

export const Container = styled.div`
  ${({ theme }) => css`
    margin-inline: auto;
    max-width: ${theme.grid.container};
    padding-inline: calc(${theme.grid.gutter} / 2);
    width: 100%;

    ${media.lessThan('medium')`
      display: flex;
      flex-direction: column;
      min-height: 100vh;
    `}
  `}
`
