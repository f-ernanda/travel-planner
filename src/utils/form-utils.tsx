import Joi from 'joi'

const fieldsValidations = {
  email: Joi.string()
    .email({ minDomainSegments: 2, tlds: { allow: ['com', 'net'] } })
    .required(),
  password: Joi.string().required()
}

export type FieldErrors = {
  [key: string]: string
}

function getFieldErrors(objError: Joi.ValidationResult) {
  const errors: FieldErrors = {}

  if (objError.error) {
    objError.error.details.forEach((err) => {
      errors[err.path.join('.')] = err.message
    })
  }

  return errors
}

type SignInValidateParams = {
  email: string
  password: string
}

export function signInValidate(values: SignInValidateParams) {
  const { email, password } = fieldsValidations
  const schema = Joi.object({ email, password })

  return getFieldErrors(schema.validate(values, { abortEarly: false }))
}
