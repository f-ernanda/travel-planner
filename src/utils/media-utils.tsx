'useClient'

import styled, { css } from 'styled-components'
import media, { DefaultBreakpoints } from 'styled-media-query'

type breakpoint = keyof DefaultBreakpoints

type MediaWrapperProps = {
  $lessThan?: breakpoint
  $greaterThan?: breakpoint
}

const mediaWrapperModifiers = {
  lessThan: (size: breakpoint) => css`
    ${media.lessThan(size)`
      display: block;
    `}
  `,

  greaterThan: (size: breakpoint) => css`
    ${media.greaterThan(size)`
      display: block;
    `}
  `
}

export const MediaWrapper = styled.div<MediaWrapperProps>`
  ${({ $lessThan, $greaterThan }) => css`
    display: none;

    ${$lessThan && mediaWrapperModifiers.lessThan($lessThan)}
    ${$greaterThan && mediaWrapperModifiers.greaterThan($greaterThan)}
  `}
`
