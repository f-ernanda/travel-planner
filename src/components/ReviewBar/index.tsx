import { Star, StarHalf, StarOutline } from '@styled-icons/material-rounded'
import * as Styled from './styles'

type ReviewBarProps = {
  $totalFilled: number
}

const ReviewBar = ({ $totalFilled }: ReviewBarProps) => {
  const totalStars = 5
  const stars = []

  for (let i = 0; i < totalStars; i++) {
    if (i < $totalFilled) {
      if (i + 0.5 < $totalFilled) {
        stars.push(<Star key={i} />)
      } else {
        stars.push(<StarHalf key={i} />)
      }
    } else {
      stars.push(<StarOutline key={i} />)
    }
  }

  return <Styled.Wrapper>{stars}</Styled.Wrapper>
}

export default ReviewBar
