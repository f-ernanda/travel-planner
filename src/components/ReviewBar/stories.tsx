import { Meta, StoryObj } from '@storybook/react'
import ReviewBar from '.'

export default {
  title: 'ReviewBar',
  component: ReviewBar,
  args: {
    $totalFilled: 2.3
  }
} as Meta

export const Default: StoryObj = {}
