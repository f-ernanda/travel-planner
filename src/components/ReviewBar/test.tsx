import { renderWithTheme } from '@/utils/test-utils'
import ReviewBar from '.'

describe('<ReviewBar />', () => {
  it('should render the matched icon', () => {
    const { container } = renderWithTheme(<ReviewBar />)

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the icon correctly with 3 filled stars', () => {
    const { container } = renderWithTheme(<ReviewBar $totalFilled={3} />)

    const filledStars = container.querySelectorAll('.star-filled')
    expect(filledStars.length).toBe(3)

    const emptyStars = container.querySelectorAll('.star-outline')
    expect(emptyStars.length).toBe(2)
  })

  it('should render the icon correctly with 2.5 filled stars', () => {
    const { container } = renderWithTheme(<ReviewBar $totalFilled={2.5} />)

    const filledStars = container.querySelectorAll('.star-filled')
    expect(filledStars.length).toBe(2)

    const halfFilledStars = container.querySelectorAll('.star-half')
    expect(halfFilledStars.length).toBe(1)

    const emptyStars = container.querySelectorAll('.star-outline')
    expect(emptyStars.length).toBe(2)
  })

  it('should render the icon correctly u0 filled stars', () => {
    const { container } = renderWithTheme(<ReviewBar $totalFilled={0} />)

    const filledStars = container.querySelectorAll('.star-filled')
    expect(filledStars.length).toBe(0)

    const emptyStars = container.querySelectorAll('.star-outline')
    expect(emptyStars.length).toBe(5)
  })
})
