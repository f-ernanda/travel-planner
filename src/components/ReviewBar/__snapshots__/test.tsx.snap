// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReviewBar /> should render the matched icon 1`] = `
.c1 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  display: flex;
  justify-content: center;
}

.c0 >* {
  color: #0D8C84;
  flex: 0 0 100%;
  max-width: 4.8rem;
}

<div
  class="c0"
>
  <svg
    aria-hidden="true"
    class="c1 star-filled"
    fill="currentColor"
    focusable="false"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="m12 17.27 4.15 2.51c.76.46 1.69-.22 1.49-1.08l-1.1-4.72 3.67-3.18c.67-.58.31-1.68-.57-1.75l-4.83-.41-1.89-4.46c-.34-.81-1.5-.81-1.84 0L9.19 8.63l-4.83.41c-.88.07-1.24 1.17-.57 1.75l3.67 3.18-1.1 4.72c-.2.86.73 1.54 1.49 1.08l4.15-2.5z"
    />
  </svg>
  <svg
    aria-hidden="true"
    class="c1 star-filled"
    fill="currentColor"
    focusable="false"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="m12 17.27 4.15 2.51c.76.46 1.69-.22 1.49-1.08l-1.1-4.72 3.67-3.18c.67-.58.31-1.68-.57-1.75l-4.83-.41-1.89-4.46c-.34-.81-1.5-.81-1.84 0L9.19 8.63l-4.83.41c-.88.07-1.24 1.17-.57 1.75l3.67 3.18-1.1 4.72c-.2.86.73 1.54 1.49 1.08l4.15-2.5z"
    />
  </svg>
  <svg
    aria-hidden="true"
    class="c1 star-filled"
    fill="currentColor"
    focusable="false"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="m12 17.27 4.15 2.51c.76.46 1.69-.22 1.49-1.08l-1.1-4.72 3.67-3.18c.67-.58.31-1.68-.57-1.75l-4.83-.41-1.89-4.46c-.34-.81-1.5-.81-1.84 0L9.19 8.63l-4.83.41c-.88.07-1.24 1.17-.57 1.75l3.67 3.18-1.1 4.72c-.2.86.73 1.54 1.49 1.08l4.15-2.5z"
    />
  </svg>
  <svg
    aria-hidden="true"
    class="c1 star-filled"
    fill="currentColor"
    focusable="false"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="m12 17.27 4.15 2.51c.76.46 1.69-.22 1.49-1.08l-1.1-4.72 3.67-3.18c.67-.58.31-1.68-.57-1.75l-4.83-.41-1.89-4.46c-.34-.81-1.5-.81-1.84 0L9.19 8.63l-4.83.41c-.88.07-1.24 1.17-.57 1.75l3.67 3.18-1.1 4.72c-.2.86.73 1.54 1.49 1.08l4.15-2.5z"
    />
  </svg>
  <svg
    aria-hidden="true"
    class="c1 star-filled"
    fill="currentColor"
    focusable="false"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="m12 17.27 4.15 2.51c.76.46 1.69-.22 1.49-1.08l-1.1-4.72 3.67-3.18c.67-.58.31-1.68-.57-1.75l-4.83-.41-1.89-4.46c-.34-.81-1.5-.81-1.84 0L9.19 8.63l-4.83.41c-.88.07-1.24 1.17-.57 1.75l3.67 3.18-1.1 4.72c-.2.86.73 1.54 1.49 1.08l4.15-2.5z"
    />
  </svg>
</div>
`;
